### PR TITLE
Handling Empty Narrows

### LINKED ISSUE #895: Crash for user with no subscriptions (eg. new bot)
Currently producing a fresh bot and using the zuliprc in ZT causes a crash, unless some streams are pre-subscribed for it.

This shouldn't be an issue for regular uses, who tend to get subscriptions automatically, but it's possible that a user could lose their subscriptions, so we should avoid this issue in any case.

### LINKED ISSUE #1506: Improve handling of narrows (views) with no messages
### Bug description

When selecting a new narrow (view) to show, via a hotkey, a button in either side panel, or using the message search (via `/`), the message focus changes to a message in the resulting view.

When the message list for that narrow/view is **empty**, ie. no messages are found, there is no visible message to gain focus, and the UI behaves strangely.

This can occur for many narrows/views, including:
- streams (channels)
  - #607 
  - #259
- searches which result in no results
  - #580
- direct messages with a specific user (eg. if not messaged them before)
  - #1480
- starred messages
  - ([#zulip-terminal > keyboard movement stops with no messages starred section](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/keyboard.20movement.20stops.20with.20no.20messages.20starred.20section))
- topics (difficult, but possible via exiting compose to a new topic)
- mentions (if never mentioned?)
- all direct messages (?)

This results in
- no obvious focused UI element (#607)
- the 'Current message recipients' not updating (remaining that of the previously focused message)
- the message narrow potentially not updating (#1480, #259)

It is possible to navigate out of this state by moving sideways and down (#580), or using one of the 'global' hotkeys such as `w` or `q`, P/#/f (though *not* `a`). However, these are workarounds.

This has been discussed variously in #zulip-terminal, and in at least the above linked issues, as well as early work on it in #278.

### What did you expect to happen?

Certainly:
- Empty the current message recipients content, or hide that section entirely (as per another issue)
- Ensure the narrow title is properly updated (eg. starred, etc.)

In addition:

1) Unlike the web app, we expect the focus to move somewhere after a new narrow is requested, so a placeholder in the message list may be the easiest short-term solution to use.

This was worked on historically in #278 as a fix for #259, where a 'fake message' was used as a placeholder. However, as demonstrated by the more complex logic we now support for searching/filtering the side panels, it should be reasonable to support a custom element for 'no results found' to be shown instead.

Even so, that solution would need to support various cases where messages 'arrive' in that empty narrow, and the placeholder is replaced by them.

2) Alternatively, focus could be retained in the previous location, if the resulting narrow is found to be empty. This would be convenient for message searches, where an empty set of results might be reasonably expected by a user to leave the focus in the search box to adjust the search.

3) Some combination of (1) and (2), since
- a placeholder shows that the narrow is empty (no messages)
- a placeholder can be focused, allowing easy movement horizontally across the app (left/right), such as from left panel to right panel.

### Notes

- I'm closing many of the linked issues, since they are essentially duplicates, ie. examples of the symptoms from the underlying problem.
- This could unblock other features, such as #1049

### PR DESCRIPTION
**Pull request by Niloth-p**
### What does this PR do, and why?
Fixes all the bugs related to empty narrows, and recipient bars when a message is not selected.

### External discussion & connections
- [ ] Discussed in **#zulip-terminal** in `topic`
- [x] Fully fixes #895
- [x] Partially fixes issue #1506 
- [x] Builds upon previous unmerged work in PR #278
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### Solution Space Exploration

Before settling on the dummy message approach, I explored a few other solutions as well:

1. I explored not switching narrows to empty narrows. We'd need to do some refactoring to `get_message_ids_in_current_narrow` to allow checking if a future narrow is empty, before switching to it.

2. And I also explored whether we could expand our system of handling narrows to also cater to narrows without messages, without handling it as an exception case and introducing a dummy message. I tried to move `top_search_bar` and `recipient_header` outside of `MessageBox` to change how they're tightly tied to the focused message, which is why they're not updated currently when in an empty narrow, and the erratic behavior.

3. Instead of a dummy message, I gave a shot at using a Text widget replacing the entire MessageView, like we do with the `PanelSearchBox`es when there are no search hits. But yes, I ran into a few cases with focus issues.
I tried moving the Text element inside, and modifying/turning off the `ModListWalker`'s action each time, to make it behave similar to the `SimpleFocusListWalker`, but that still wasn't a clean fix.

And other similar tangents, before realising the dummy message approach is the solid one, even if we were to implement any of these in the future, especially the 1st approach, the dummy message should be a backup.

I'm sure a more experienced person would have been able to figure this out much faster than I did, and without all this exploration, so I'd be interested in hearing the thinking process of someone who was able to / can narrow in on (pun intended) the dummy message solution right away.

I also decided against removing the message recipients section for empty narrows, because currently when opening the app, it shows up with "DONT HIDE", so I went with a solution that conforms to that behavior.

### Outstanding aspect(s)
- I've added the features for the dummy message, before adding the dummy message, so that it starts working as soon as added.
- Placed the current narrow state tracking variable in the controller class.
- I wasn't able to use even a single line of code as-is from #278, as this PR is quite different and uses several different decisions, so I just decided to add 278's author as a co-author for all ideas that were inspired from their PR.
- I haven't added "bugfix" prefix to the commit messages, should I add it to the commit that wires the dummy message?

### Next Steps
- [ ] Adding tests for empty narrows. None of the commits currently have any new tests.
- [ ] This introduces a bug. When you have an empty search narrow. And you receive a message event for that narrow, the message does not display until you restart the application, although it adds to the unread count.
I didn't want to hold back on the reviews any longer, so I've pushed the PR and will look into this soon.

### Follow-ups
- [ ] When activating user buttons, check if there exist previous DMs with the user, and if not, just open the compose box and do not switch to that narrow.
- [ ] When a new message arrives in an empty DM narrow, the recipient header (above the message, not the one in the recipient bar) is not inserted newly. It automatically gets fixed when one more new message arrives. I think it's not a high priority bug, since the recipient bar already matches the recipient header for the DM case.

### How did you test this?
- [x] Manually - Behavioral changes
- [x] Manually - Visual changes
- [x] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

### CODE REVIEW

**Review by neiljp — COMMENTED** [review]
@Niloth-p I know I replied in the channel, but wanted to follow up here here too.

Broadly this looks great, though there are queries over which specific parts we cover, eg. which keys dummy messages respond to. There are also possibly-missing aspects to consider sooner or later, such as how and when to narrow from a dummy message.

I think each specific feedback comment should be easy to follow, but let me know if not. 

I'm increasingly thinking the empty-narrow might be best tracked in the model along with the 'current narrow' and associated data?

I think I addressed your outstanding aspects, but I'd add it only to the last one, and maybe only when it's close to ready to avoid spamming the issue ;) (see a recent czo topic!)

Oh, one thing I did find that I didn't note elsewhere is that the DM area doesn't indicate *which* recipients you're chatting with. This is a side-effect of how we handle the two sections at the top, but makes me wonder if the current-message-recipients should show the *intended* narrow/recipients, even if there are none. In a comment I wondered about treating the recipients as blank, but this would be clearer? Re DMs specifically, if I do enter in the user list to someone new, it usefully makes it empty, but as soon as I exit the user it treats it as some kind of generic narrow - `x` doesn't resume that user (which isn't indicated). I'm not sure what happens in other narrows?

File: zulipterminal/ui_tools/utils.py [hunk 1]
```diff
@@ -29,6 +29,46 @@ def create_msg_box_list(
     focus_msg = None
     last_msg = last_message
     muted_msgs = 0  # No of messages that are muted.
+
+    # Add a dummy message if no new or old messages are found
+    if not message_list and not last_msg:
+        message_type = "stream" if model.stream_id is not None else "private"
+        dummy_message = {
```
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/utils.py:36]: To clarify
- I know that the function in this file is used a lot and it makes it simpler to centralize the calling point here
- passing in a complex 'fake' message that has to have all the values to render properly seems more complicated than saying 'please make this a dummy with these parameters'?
